torch/csrc/api/include/torch/data/dataloader.h

```diff
@@ -0,0 +1,233 @@
+  }
+
+  template <typename LoopFunction>
+  void loop(LoopFunction function) {
+    while (auto maybe_batch = next()) {
+      function(*maybe_batch);
+    }
+    reset();
+  }
+
+  void join() {
+    if (joined_) {
+      return;
+    }
+    shuttle_.drain();
+    // Send one 'quit' message per worker. Since a worker dies (exits its
+    // thread) after receiving this message, each `QuitWorker()` message will be
+    // read by exactly one worker.
+    for (size_t w = 0; w < options_.workers_; ++w) {
+      push_job(QuitWorker());
+    }
+    for (auto& worker : workers_) {
+      worker.join();
+    }
+    joined_ = true;
+  }
+
+  const DataLoaderOptions& options() const noexcept {
+    return options_;
+  }
+
+  size_t dataset_size() const noexcept {
+    return dataset_size_;
+  }
+
+ private:
+  struct Sequenced {
+    Sequenced() = default;
+    Sequenced(size_t sqn) : sequence_number(sqn) {}
+    size_t sequence_number;
+  };
+
+  struct QuitWorker {};
+  // Job = Enum(QuitWorker, IndexBatch)
+  struct Job : Sequenced {
+    Job() = default;
+    Job(QuitWorker q, size_t sqn) : Sequenced(sqn), quit(q) {}
+    Job(IndexBatch&& i, size_t sqn)
+        : Sequenced(sqn), index_batch(std::move(i)) {}
+    at::optional<QuitWorker> quit;
+    at::optional<IndexBatch> index_batch;
+  };
+
+  struct Result : Sequenced {
+    Result() = default;
+    Result(Batch&& b, size_t sqn) : Sequenced(sqn), batch(std::move(b)) {}
+    Batch batch;
+  };
+
+  void worker_thread(DatasetType dataset) {
+    while (true) {
+      auto job = shuttle_.pop_job();
+      if (job.quit) {
+        break;
+      }
+      auto batch = dataset.batch(std::move(*job.index_batch));
+      shuttle_.push_result({std::move(batch), job.sequence_number});
+    }
+  }
+
+  at::optional<IndexBatch> get_index_batch() {
+    auto indices = sampler_.next(options_.batch_size_);
+    if (!indices ||
+        (indices->size() < options_.batch_size_ && options_.drop_last_)) {
+      return at::nullopt;
+    }
+    AT_ASSERT(!indices->empty());
+    return indices;
+  }
+
+  template <typename T>
+  void push_job(T&& value) {
+    shuttle_.push_job({std::forward<T>(value), sequence_number_++});
+  }
+
+  const DataLoaderOptions options_;
+
+  at::optional<DatasetType> dataset_;
+  size_t dataset_size_;
+
+  SamplerType sampler_;
+  size_t sequence_number_{0};
+  std::vector<std::thread> workers_;
+  detail::DataShuttle<Job, Result> shuttle_;
+  std::unique_ptr<detail::sequencers::Sequencer<Result>> sequencer_;
+  bool joined_ = false;
+};
+
+template <typename DatasetType, typename SamplerType>
+std::unique_ptr<DataLoader<DatasetType, SamplerType>> data_loader(
+    DatasetType dataset,
+    DataLoaderOptions options,
+    SamplerType sampler) {
+  return torch::make_unique<DataLoader<DatasetType, SamplerType>>(
+      std::move(dataset), std::move(options), std::move(sampler));
+}
+
+template <
+    typename DatasetType,
+    typename SamplerType = samplers::RandomSampler<>,
```

Comment:
In that case it can't be defaulted. Also `DatasetType` can't be defaulted or inferred inside the class (only in an external function like `data_loader(...)`, so you'd have to specify it either way. Not sure about the gain here